File: apps/api/app/models/integration.py
import uuid
from datetime import datetime

from sqlalchemy import Boolean, DateTime, ForeignKey, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base


class Integration(Base):
    __tablename__ = "integrations"

    id: Mapped[str] = mapped_column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id: Mapped[str] = mapped_column(ForeignKey("users.id"), nullable=False, index=True)
    provider: Mapped[str] = mapped_column(String, nullable=False)
    status: Mapped[str] = mapped_column(String, default="disconnected", nullable=False)
    connected_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    last_sync_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    error_message: Mapped[str | None] = mapped_column(Text)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )

    user: Mapped["User"] = relationship(back_populates="integrations")  # noqa: F821


class IntegrationSettings(Base):
    __tablename__ = "integration_settings"

    id: Mapped[str] = mapped_column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id: Mapped[str] = mapped_column(ForeignKey("users.id"), nullable=False, unique=True, index=True)
    slack_enabled: Mapped[bool] = mapped_column(Boolean, default=False)
    slack_channel_id: Mapped[str | None] = mapped_column(String)
    gmail_enabled: Mapped[bool] = mapped_column(Boolean, default=False)
    linear_enabled: Mapped[bool] = mapped_column(Boolean, default=False)
    jira_enabled: Mapped[bool] = mapped_column(Boolean, default=False)
    asana_enabled: Mapped[bool] = mapped_column(Boolean, default=False)
    hubspot_enabled: Mapped[bool] = mapped_column(Boolean, default=False)
    salesforce_enabled: Mapped[bool] = mapped_column(Boolean, default=False)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )

File: packages/shared/src/types/integration.ts
export type IntegrationProvider =
  | "slack"
  | "gmail"
  | "linear"
  | "jira"
  | "asana"
  | "hubspot"
  | "salesforce"
  | "google_calendar";

export type IntegrationStatus = "connected" | "disconnected" | "error";

export interface Integration {
  id: string;
  userId: string;
  provider: IntegrationProvider;
  status: IntegrationStatus;
  connectedAt?: string;
  lastSyncAt?: string;
  errorMessage?: string;
}

export interface IntegrationSettings {
  userId: string;
  slackEnabled: boolean;
  slackChannelId?: string;
  gmailEnabled: boolean;
  linearEnabled: boolean;
  jiraEnabled: boolean;
  asanaEnabled: boolean;
  hubspotEnabled: boolean;
  salesforceEnabled: boolean;
}
